### PR TITLE
Extract updates from all game objects to send over network

### PR DESCRIPTION
**Pull request by SourenP**
- Move Group and Mine update logic into their respective classes
- Move Group and Mine into the common dir
- Move `is_active` into update instead of inferring from set of clients
- Extract updates from all game objects on server side
- Create `ThreadSafeData` - just a class that has a single var `data` that is thread safe to read from/write to. This is used to read/write game state in the networking code.

### CODE REVIEW

File: src/client/Client.cpp [hunk 1]
```diff
@@ -50,144 +56,49 @@ void Client::draw(sf::RenderTarget& target, sf::Shader* shader, bool use_shader)
...
-        int client_mine_id = mMineIdToClientMine[mine_update.mine_id];
-        ClientMine* client_mine = mClientMines[client_mine_id];
+    // Send updates
+    mNetworkingClient->setClientTCPUpdate(mClientTCPUpdate);
+    mNetworkingClient->setClientUDPUpdate(mClientUDPUpdate);
 
-        client_mine->setPosition(
-                sf::Vector2f(mine_update.x_pos, mine_update.y_pos));
-        client_mine->getCircle()->setRadius(mine_update.radius);
+    // Apply updates
```
Comment (copacetic) [inline src/client/Client.cpp]: delete useless comment

File: src/client/Client.cpp [hunk 2]
```diff
@@ -50,144 +56,49 @@ void Client::draw(sf::RenderTarget& target, sf::Shader* shader, bool use_shader)
...
-}
-
-
-sf::Uint32 Client::getId() const {
-    return mId;
 }
 
 void Client::handleEvents(sf::Event& event) {
     if (event.type == sf::Event::KeyPressed) {
+        // Groupable
```
Comment (copacetic) [inline src/client/Client.cpp]: delete

File: src/client/Client.cpp [hunk 3]
```diff
@@ -50,144 +56,49 @@ void Client::draw(sf::RenderTarget& target, sf::Shader* shader, bool use_shader)
...
         if (sf::Keyboard::isKeyPressed(mKeys.group)) {
-            mGroupable ^= true;
+            mClientTCPUpdate.groupable ^= true;
             return;
         }
-        mDirection = sf::Vector2f(0.f, 0.f);
+
+        // Direction
```
Comment (copacetic) [inline src/client/Client.cpp]: delete

File: src/common/ThreadSafeVector.hpp [hunk 4]
```diff
@@ -33,6 +33,12 @@ class ThreadSafeVector {
             mVector.push_back(element);
             unlock();
         }
+        std::vector<T> forceGetAndClear() {
```
Comment (copacetic) [inline src/common/ThreadSafeVector.hpp]: use forceGet and forceClear --souren

File: src/server/GameController.cpp [hunk 5]
```diff
@@ -17,58 +17,51 @@ GameController::GameController(size_t max_player_count, size_t max_mine_count):
...
 }
 
-/**
-    Updates the active status and  properties of players based on updates recieved from the clients.
-*/
-void GameController::updatePlayers(const client_inputs& cis) {
+void GameController::updatePlayers(const ClientInputs& cis) {
     for (const auto& new_client_id : cis.new_client_ids) {
-        mClientToPlayer[new_client_id] = createPlayerWithGroup();
+        int new_player_id = mClientToPlayer[new_client_id] = createPlayerWithGroup();
```
Comment (copacetic) [inline src/server/GameController.cpp]: fancyboi

File: src/server/NetworkingServer.cpp [hunk 6]
```diff
@@ -164,30 +166,32 @@ void NetworkingServer::handleApiCommand(sf::Socket::Status status, sf::Packet co
...
-        mClientGroupable.set(client_id, groupable);
-    }
+    packet << api_command << pi;
+    client.send(packet);
+}
+
+void NetworkingServer::setClientTCPUpdate(sf::Packet packet, int client_id) {
+    ClientTCPUpdate client_tcp_update;
+    packet >> client_tcp_update;
+    ClientIdAndTCPUpdate client_id_and_tcp_update = {client_id, client_tcp_update};
```
Comment (copacetic) [inline src/server/NetworkingServer.cpp]: bruhh im gonna puke what is this struct

File: src/server/NetworkingServer.cpp [hunk 7]
```diff
@@ -203,84 +207,51 @@ void NetworkingServer::registerClient(sf::TcpSocket& client) {
...
-    std::vector<client_direction_update> client_direction_updates;
-    for (const auto& client_move : mClientMoves.forceCopy()) {
-        client_direction_update cd = {
-            client_move.first, client_move.second.x, client_move.second.y};
-        client_direction_updates.push_back(cd);
-    }
-
-    return client_direction_updates;
+void NetworkingServer::setClientToPlayerId(sf::Int32 client_id, int player_id) {
+    mClientToPlayerIds.forceSet(client_id, player_id);
```
Comment (copacetic) [inline src/server/NetworkingServer.cpp]: why isn't client id == player id?

File: src/client/NetworkingClient.cpp [hunk 8]
```diff
@@ -79,90 +79,97 @@ void NetworkingClient::ReadRegistrationResponse() {
...
         ApiCommand api_command;
-
         if (mApiClient->receive(api_response) == sf::Socket::Done) {
-            if (api_response >> api_command &&
-                api_command.command == (sf::Uint32)APICommandType::toggle_groupable) {
-                // this is a noop, just a place we can add code for when the server acknowledges
-                // it received our groupability toggle
+            api_response >> api_command;
+            if (api_command.command == (sf::Uint32) APICommandType::player_id) {
+                PlayerId pi;
```
[the remark this answers is not in the document]
Comment (SourenP) [reply src/client/NetworkingClient.cpp]: no :<

File: src/server/NetworkingServer.cpp (hunk 6 above)
Comment (SourenP) [reply src/server/NetworkingServer.cpp]: go ahead ;)

File: src/server/NetworkingServer.cpp (hunk 7 above)
Comment (SourenP) [reply src/server/NetworkingServer.cpp]: created #61

**Pull request merged**